File: code_snippets/java/src/main/java/usecases/eventprocessing/UserUpdatesService.java
package usecases.eventprocessing;

import static usecases.utils.ExampleStubs.*;

import dev.restate.sdk.JsonSerdes;
import dev.restate.sdk.ObjectContext;
import dev.restate.sdk.annotation.Handler;
import dev.restate.sdk.annotation.VirtualObject;
import java.time.Duration;
import usecases.utils.UserUpdate;

// <start_here>
@VirtualObject
public class UserUpdatesService {

  // <mark_1>
  @Handler
  public void updateUserEvent(ObjectContext ctx, UserUpdate event) {
    // </mark_1>
    // <mark_3>
    String userId = ctx.run(JsonSerdes.STRING, () -> updateUserProfile(event.getProfile()));
    // </mark_3>

    // <mark_4>
    while (userId.equals("NOT_READY")) {
      // <mark_2>
      ctx.sleep(Duration.ofMillis(5000));
      // </mark_2>
      // <mark_3>
      userId = ctx.run(JsonSerdes.STRING, () -> updateUserProfile(event.getProfile()));
      // </mark_3>
    }

    String finalUserId = userId;
    // <mark_3>
    String roleId =
        ctx.run(JsonSerdes.STRING, () -> setUserPermissions(finalUserId, event.getPermissions()));
    ctx.run(() -> provisionResources(finalUserId, roleId, event.getResources()));
    // </mark_3>
    // </mark_4>
  }
}
// <end_here>

File: code_snippets/kotlin/src/main/kotlin/usecases/eventprocessing/UserUpdatesService.kt
package usecases.eventprocessing

import dev.restate.sdk.annotation.Handler
import dev.restate.sdk.annotation.VirtualObject
import dev.restate.sdk.kotlin.ObjectContext
import dev.restate.sdk.kotlin.runBlock
import kotlin.time.Duration.Companion.milliseconds

// <start_here>
@VirtualObject
class UserUpdatesService {
  // <mark_1>
  @Handler
  suspend fun updateUserEvent(ctx: ObjectContext, event: UserUpdate) {
    // </mark_1>
    // <mark_3>
    var userId = ctx.runBlock { updateUserProfile(event.profile) }

    // </mark_3>

    // <mark_4>
    while (userId == "NOT_READY") {
      // <mark_2>
      ctx.sleep(5000.milliseconds)
      // </mark_2>
      // <mark_3>
      userId = ctx.runBlock { updateUserProfile(event.profile) }
      // </mark_3>
    }

    val finalUserId = userId
    // <mark_3>
    val roleId = ctx.runBlock { setUserPermissions(finalUserId, event.permissions) }
    ctx.runBlock { provisionResources(finalUserId, roleId, event.resources) }
    // </mark_3>
  }
  // </mark_4>
}
// <end_here>

class UserUpdate(var profile: String, var permissions: String, var resources: String)

fun updateUserProfile(profile: String): String {
  return if (Math.random() < 0.8) "NOT_READY" else "$profile-id"
}

fun setUserPermissions(userId: String?, permissions: String): String {
  return permissions
}

fun provisionResources(userId: String?, role: String?, resources: String?): String {
  return ""
}
